src/protocol/maliciously_secure_mul.rs

```diff
@@ -0,0 +1,346 @@
+
+impl AsRef<str> for Step {
+    fn as_ref(&self) -> &str {
+        match self {
+            Self::DuplicateMultiply => "duplicate_multiply",
+            Self::RandomnessForValidation => "randomness_for_validation",
+        }
+    }
+}
+
+///
+/// Implementation drawn from:
+/// "Fast Large-Scale Honest-Majority MPC for Malicious Adversaries"
+/// by by K. Chida, D. Genkin, K. Hamada, D. Ikarashi, R. Kikuchi, Y. Lindell, and A. Nof
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// Protocol 5.3 "Computing Arithmetic Circuits Over Any Finite F"
+/// Step 5: "Circuit Emulation"
+/// (In our case, simplified slightly because δ=1)
+/// When `G_k` is a multiplication gate:
+/// Given tuples:  `([x], [r · x])` and `([y], [r · y])`
+/// (a) The parties call `F_mult` on `[x]` and `[y]` to receive `[x · y]`
+/// (b) The parties call `F_mult` on `[r · x]` and `[y]` to receive `[r · x · y]`.
+///
+/// As each multiplication gate affects Step 6: "Verification Stage", the Security Validator
+/// must be provided. The two outputs of the multiplication, `[x · y]` and  `[r · x · y]`
+/// will be provided to this Security Validator, and will update two information-theoretic MACs.
+///
+/// It's cricital that the functionality `F_mult` is secure up to an additive attack.
+/// `SecureMult` is an implementation of the IKHC multiplication protocol, which has this property.
+///
+pub struct MaliciouslySecureMul<'a, N, F> {
+    ctx: &'a ProtocolContext<'a, N>,
+    record_id: RecordId,
+    security_validator: Arc<Mutex<SecurityValidator<F>>>,
+}
+
+impl<'a, N: Network, F: Field> MaliciouslySecureMul<'a, N, F> {
+    #[must_use]
+    pub fn new(
+        ctx: &'a ProtocolContext<'a, N>,
+        record_id: RecordId,
+        security_validator: Arc<Mutex<SecurityValidator<F>>>,
+    ) -> Self {
+        Self {
+            ctx,
+            record_id,
+            security_validator,
+        }
+    }
+
+    /// Executes two parallel multiplications;
+    /// `A * B`, and `rA * B`, yielding both `AB` and `rAB`
+    /// both `AB` and `rAB` are provided to the security validator
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    /// ## Panics
+    /// Panics if the mutex is found to be poisoned
+    pub async fn execute(
+        self,
+        a: MaliciousReplicated<F>,
+        b: MaliciousReplicated<F>,
+    ) -> Result<MaliciousReplicated<F>, BoxError> {
+        let duplicate_ctx = self.ctx.narrow(&Step::DuplicateMultiply);
+        let (ab, rab) = try_join(
+            SecureMul::new(self.ctx, self.record_id).execute(a.x(), b.x()),
+            SecureMul::new(&duplicate_ctx, self.record_id).execute(a.rx(), b.x()),
+        )
+        .await?;
+
+        let malicious_ab = MaliciousReplicated::new(ab, rab);
+
+        self.security_validator.lock().unwrap().accumulate_macs(
```

Comment:
this will make all multiplications to execute serially one after the other. I wonder if `SecurityValidator` should just be `Send` and provide `accumulate_macs` method that takes `&self`. We could back this implementation by a channel. I wonder how crazy it sounds to implement fields backed by atomic value rather than just an integer...

Reply:
I don't think that this serializes multiplications if this action is short.  It does block, but a short bit of non-concurrent code might be OK.  I might prefer to hide the implementation details (mutex vs. channel) under an API though.  That means creating a wrapper that either holds the producer end of an mpsc channel or the `Arc<Mutex<X>>`.